[FLINK-29879][filesystem] introduce operators for merging files in batch mode.

luoyuxia opened the pull request:
<!--
*Thank you very much for contributing to Apache Flink - we are happy that you want to help us improve Flink. To help the community review your contribution in the best possible way, please go through the checklist below, which will get the contribution into a shape in which it can be best reviewed.*

*Please understand that we do not do this to make contributions to Flink a hassle. In order to uphold a high standard of quality for code contributions, while at the same time managing a large number of contributions, we need contributors to prepare the contributions well, and give reviewers enough contextual information for the review. Please also understand that contributions that do not follow this guide will take longer to review and thus typically be picked up with lower priority by the community.*

## Contribution Checklist

  - Make sure that the pull request corresponds to a [JIRA issue](https://issues.apache.org/jira/projects/FLINK/issues). Exceptions are made for typos in JavaDoc or documentation files, which need no JIRA issue.
  
  - Name the pull request in the form "[FLINK-XXXX] [component] Title of the pull request", where *FLINK-XXXX* should be replaced by the actual issue number. Skip *component* if you are unsure about which is the best component.
  Typo fixes that have no associated JIRA issue should be named following this pattern: `[hotfix] [docs] Fix typo in event time introduction` or `[hotfix] [javadocs] Expand JavaDoc for PuncuatedWatermarkGenerator`.

  - Fill out the template below to describe the changes contributed by the pull request. That will give reviewers the context they need to do the review.
  
  - Make sure that the change passes the automated tests, i.e., `mvn clean verify` passes. You can set up Azure Pipelines CI to do that following [this guide](https://cwiki.apache.org/confluence/display/FLINK/Azure+Pipelines#AzurePipelines-Tutorial:SettingupAzurePipelinesforaforkoftheFlinkrepository).

  - Each pull request should address only one issue, not mix up code from multiple issues.
  
  - Each commit in the pull request has a meaningful commit message (including the JIRA id)

  - Once all items of the checklist are addressed, remove the above text and this checklist, leaving only the filled out template below.


**(The sections below can be removed for hotfixes of typos)**
-->

## What is the purpose of the change
Introduce operators for merging files in batch mode. 
The whole pipeline will be `BatchFileWriter ` -> `BatchCompactCoordinator ` -> `BatchCompactOperator ` -> `BatchPartitionCommitterSink`


## Brief change log
  -  Introuce `BatchFileWriter` which will write files and emit the written path to downstream. The logic for writing is similar to `FileSystemOutputFormat`.
  -  introuce `BatchCompactCoordinator `which will collect all written files and decide what files should be merged to a single file. The main logic is similar to `CompactCoordinator`
  -  Introuce `BatchCompactOperator` which will do the actual compaction for files, the main logic is similar to `CompactOperator`
  -  Introuce `BatchPartitionCommitterSink` which will commit files and partitions, the logic for committing is similar to `FileSystemOutputFormat`


## Verifying this change
Added UT for each single operator

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / **not documented**)

luoyuxia commented:
@flinkbot run azure

Aitozi reviewed (commented):
Hi @luoyuxia , I left some comments for this PR.

Aitozi commented on flink-connectors/flink-connector-files/src/main/java/org/apache/flink/connector/file/table/PartitionLoader.java:
```diff
@@ -148,23 +150,35 @@ private void overwrite(Path destDir) throws Exception {
-                            fs.rename(srcPath, destPath);
-                        } else {
-                            FileUtils.copy(srcPath, destPath, true);
+    private void moveFiles(List<Path> srcPaths, Path destDir, boolean srcPathIsDir)
```
Can we check the srcPaths is dir by `fs.getFileStatus(srcDir).isDir()` here, If it's dir, then list the directory

luoyuxia replied:
We can check by `fs.getFileStatus(srcDir).isDir()`, but I prefer to still keep the logic in here. The reason is as follows:
1: I think it's more clear to use method argument to make sure these srcpath are files. Also, if we check the path is dir or not. Do we still need to check the children path is dir or not?
2: It'll cost time to list file status

Aitozi commented on flink-connectors/flink-connector-files/src/main/java/org/apache/flink/connector/file/table/batch/compact/BatchCompactCoordinator.java:
```diff
@@ -0,0 +1,155 @@
+            inputFiles
+                    .computeIfAbsent(file.getPartition(), k -> new ArrayList<>())
+                    .add(file.getFile());
+        }
```
throw an exception for other unexpected type

luoyuxia replied:
done

Aitozi commented on flink-connectors/flink-connector-files/src/main/java/org/apache/flink/connector/file/table/batch/compact/BatchCompactCoordinator.java:
```diff
@@ -0,0 +1,155 @@
+
+    private static class AverageSize {
+        private final long totalSz;
+        private final int numFiles;
```
average size can be calculated eagerly

luoyuxia replied:
I don't think it's bad to calculate it later. Also, if the numFiles is zero, we can't calculate the average size.

Aitozi commented on flink-connectors/flink-connector-files/src/main/java/org/apache/flink/connector/file/table/batch/compact/BatchCompactOperator.java:
```diff
@@ -0,0 +1,114 @@
+    }
+
+    @Override
+    public void initializeState(StateInitializationContext context) throws Exception {
```
will this `initializeState`  be invoked by streaming job? BTW, I think this should be put in `open()`

luoyuxia replied:
Yes. I think put it in `initializeState` is more consistent to existing code. e.g, CompactCoordinator also init in `initializeState` method.

Aitozi commented on flink-connectors/flink-connector-files/src/main/java/org/apache/flink/connector/file/table/batch/compact/BatchCompactCoordinator.java:
```diff
@@ -0,0 +1,155 @@
+
+    @Override
+    public void initializeState(StateInitializationContext context) throws Exception {
+        super.initializeState(context);
```
will this `initializeState` be invoked by streaming job? BTW, I think this should be put in `open()`?

luoyuxia replied:
dito.

Aitozi commented on flink-connectors/flink-connector-files/src/main/java/org/apache/flink/connector/file/table/batch/compact/BatchCompactOperator.java:
```diff
@@ -0,0 +1,114 @@
+            if (path != null) {
+                compactedFiles.computeIfAbsent(partition, k -> new ArrayList<>()).add(path);
+            }
+        }
```
throw an unexpected type here

luoyuxia replied:
done

Aitozi commented on flink-connectors/flink-connector-files/src/main/java/org/apache/flink/connector/file/table/batch/compact/BatchPartitionCommitter.java:
```diff
@@ -0,0 +1,141 @@
+    @Override
+    public void initializeState(StateInitializationContext context) throws Exception {
+        super.initializeState(context);
+        partitionsFiles = new HashMap<>();
```
I think this should be put to `open()`

luoyuxia replied:
dito

beyond1920 commented on flink-connectors/flink-connector-files/src/main/java/org/apache/flink/connector/file/table/batch/compact/BatchFileWriter.java:
```diff
@@ -0,0 +1,115 @@
+                            tmpPath,
+                            getRuntimeContext().getIndexOfThisSubtask(),
+                            getRuntimeContext().getAttemptNumber(),
+                            outputFileConfig);
```
It's not safe if the job enables speculative execution because multiple attempts would write to the same file.

luoyuxia replied:
If the attemptNumber is different, the file to be written is different.

beyond1920 reviewed (commented):
@luoyuxia Thanks for contribution.
Wrong result might happen if a job enables speculative execution because operators (e.g BatchFileWriter) does not take it into consideration.
We could choose not to support speculative execution in the first version. 
However we should check parameters and reminder users this point to avoid wrong results.

beyond1920 commented on flink-connectors/flink-connector-files/src/main/java/org/apache/flink/connector/file/table/CompactFileUtils.java:
```diff
@@ -0,0 +1,158 @@
+
+        Path target = createCompactedFile(paths);
+        if (fileSystem.exists(target)) {
+            return target;
```
I'm not sure if this behavior is correct if enable speculative execution.
If one compact operator is slow, the scheduler might deploy a new attempt for it.
The compact operator in new attempt would be return directly here because the target path is already exists. But it does not do anything yet.

luoyuxia replied:
Now, I modify the logic for the generating target path . Different attempts will have different target path, so that such case won't happen then.

luoyuxia replied:
A bit of more explanation in here:
if the compact unit only contains single file, we will do nothing and make the downstream operator commit(rename) the files.

If the compact unit contains multiple files, we will read these files and write them to a target path, but the target path is different for different attempt.

So, in these operator, speculative execution won't make  any difference.

luoyuxia commented:
@beyond1920 Thanks for reviewing. I prefer to append a commit with tests to mock  the speculative execution to validate these operators.

luoyuxia commented:
> @beyond1920 Thanks for reviewing. I prefer to append a commit with tests to mock the speculative execution to validate these operators.

Actaully, I don't think we need to add tests to mock peculative execution for the test will be just like create multiple operators and make operator to process element and make sure it'll generate different paths for different attempts.
Do you think we still need to add these test. @beyond1920

Tartarus0zm commented on flink-connectors/flink-connector-files/src/main/java/org/apache/flink/connector/file/table/batch/compact/BatchFileWriter.java:
```diff
@@ -0,0 +1,115 @@
+        this.staticPartitions = staticPartitions;
+        this.formatFactory = formatFactory;
+        this.computer = computer;
+        this.outputFileConfig = outputFileConfig;
```
we should add `setChainingStrategy(ChainingStrategy.ALWAYS);`
then the Operator will be chained with upstream.

beyond1920 commented on flink-connectors/flink-connector-files/src/main/java/org/apache/flink/connector/file/table/batch/compact/BatchPartitionCommitOutputFormat.java:
```diff
@@ -0,0 +1,148 @@
+                            identifier,
+                            staticPartitions,
+                            policies);
+            committer.commitPartitionsWithFiles(partitionsFiles);
```
I'm not sure whether we could get right value of `partitionsFiles` when run `finalizeGlobal` in JobMaster.

luoyuxia replied:
Yes, you're right. Now, i replace it with `BatchPartitionCommitterSink`.

beyond1920 reviewed (commented):
After discussion with @luoyuxia offline, we would not support speculative execution in the this version. And we would continue on this work after [FLIP-281+Sink+Supports+Speculative+Execution+For+Batch+Job](https://cwiki.apache.org/confluence/display/FLINK/FLIP-281+Sink+Supports+Speculative+Execution+For+Batch+Job) is done.

luoyuxia commented:
> After discussion with @luoyuxia offline, we would not support speculative execution in the this version. And we would continue on this work after [FLIP-281+Sink+Supports+Speculative+Execution+For+Batch+Job](https://cwiki.apache.org/confluence/display/FLINK/FLIP-281+Sink+Supports+Speculative+Execution+For+Batch+Job) is done. Please fire a JIRA to issue the future work.

Yes. We can continue this after [FLIP-281+Sink+Supports+Speculative+Execution+For+Batch+Job](https://cwiki.apache.org/confluence/display/FLINK/FLIP-281+Sink+Supports+Speculative+Execution+For+Batch+Job).

But I don't think we need any issue the future work. For current design, among these four operatos, the only concern is the `BatchPartitionCommitterSink` in Speculative Execution.  In Speculative Execution, it may cause to commit files for concurrently which is problematic. But as FLIP-281 said, if sink function doesn't implement `SupportsConcurrentExecutionAttempts`, speculative execution will be disabled. So, if the FLIP-281, we don't need to do any extra thing for the `BatchPartitionCommitterSink` will disable speculative execution by default.
I may miss something, please correct me if I'm wrong.

beyond1920 commented:
If you plan to keep the changes of BufferWriters and Compact operators for speculative execution in this pr, there is no need to create a JIRA to issue this.

luoyuxia commented on flink-connectors/flink-connector-files/src/main/java/org/apache/flink/connector/file/table/CompactFileUtils.java:
```diff
@@ -0,0 +1,158 @@
+        }
+
+        Map<Path, Long> targetMap = new HashMap<>();
+        targetMap.put(target, fileSystem.getFileStatus(target).getLen());
```
It's need to log info for compaction.

luoyuxia commented:
@beyond1920 @Aitozi @Tartarus0zm Thanks for you guys reviewing. I have addressed your comments.

luoyuxia commented on flink-connectors/flink-connector-files/src/main/java/org/apache/flink/connector/file/table/batch/compact/BatchCompactOperator.java:
```diff
@@ -0,0 +1,145 @@
+ * emit the file to downstream. Also, the name of the files to be compacted is not a hidden file,
+ * it's expected these files are in a hidden or temporary directory. so please make sure
+ */
+public class BatchCompactOperator<T> extends AbstractStreamOperator<CompactOutput>
```
You can find how to do decide the parallesm in here https://github.com/apache/flink/pull/21059/files#diff-a19ed73881a2c9272de83b1b916c4b549f7ec2734202113a0659128c479e91cfR579

beyond1920 replied:
It's a little hack, is there any elegant way? cc @zhuzhurk

luoyuxia replied:
In the current framework for aqe,  I don't think we have an elegant way to decide for parallesm for this operator unless we can also decide parallesm by operator self in aqe.

luoyuxia replied:
Yes, that's true. I have created FLINK-30695 to track it. May @zhuzhurk @wanglijie95 have a look about FLINK-30695.

luoyuxia replied:
Btw, auto compact is not enabeld by default, and user can set the parallelism manually while it maybe not a good user experience.

luoyuxia commented on flink-connectors/flink-connector-files/src/main/java/org/apache/flink/connector/file/table/PartitionTempFileManager.java:
```diff
@@ -75,6 +76,19 @@ public class PartitionTempFileManager {
+                String.format(
+                        "%s%d-%s%d", TASK_DIR_PREFIX, taskNumber, ATTEMPT_PREFIX, attemptNumber);
+        this.taskTmpDir = new Path(tmpPath, taskTmpDirName);
+    }
```
done

beyond1920 commented on flink-connectors/flink-connector-files/src/main/java/org/apache/flink/connector/file/table/batch/compact/CompactFileUtils.java:
```diff
@@ -0,0 +1,119 @@
+            return null;
+        }
+
+        if (fileSystem.exists(target)) {
```
`BatchCompactOperator` call this function, when would `fileSystem.exists(target)` return true?
After restart from a failover? If yes, maybe we could clean the target path instead of directly return path?

luoyuxia replied:
In current implementation for `BatchCompactOperator` , since the target path will always a different path whenever if's fail over or differenet attempts in speculative execution.  I perfer to keep the logic in CompactFileUitls since it's like a general method without any other assumption and it's designed to return the target path directly if the target path exists.

lsyldliu commented on flink-connectors/flink-connector-files/src/main/java/org/apache/flink/connector/file/table/batch/compact/BatchPartitionCommitterSink.java:
```diff
@@ -0,0 +1,139 @@
+ * Committer operator for partition in batch mode. This is the single (non-parallel) task. It
+ * collects all the partition information including partition and written files send from upstream.
+ */
+public class BatchPartitionCommitterSink extends RichSinkFunction<CompactOutput> {
```
Whether can we extends the `AbstractStreamOperator` directly like other operators?

luoyuxia replied:
No,  Flink will expect return a `sink`.  So, here we make it extend RichSinkFunction.

lsyldliu reviewed (commented):
@luoyuxia Thanks for contribution, I left some comments.

luoyuxia commented on flink-connectors/flink-connector-files/src/main/java/org/apache/flink/connector/file/table/PartitionWriter.java:
```diff
@@ -62,4 +62,24 @@ OutputFormat<T> createNewOutputFormat(Path path) throws IOException {
+         * @param partition The partition for the newly created file.
+         * @param file The newly created file.
+         */
+        void onFileOpen(String partition, Path file);
```
Actually it should be `onFileOpen`. but I change the doc for the method, 
```java
Note that this does not mean that the file has been created in the file system. It is  only created logically and the actual file will be generated after it is committed.
```

luoyuxia commented on flink-connectors/flink-connector-files/src/main/java/org/apache/flink/connector/file/table/batch/compact/BatchFileWriter.java:
```diff
@@ -0,0 +1,124 @@
+ * An operator for writing files in batch mode. Once creating a new file to write, the writing
+ * operator will emit the written file to downstream.
+ */
+public class BatchFileWriter<T> extends AbstractStreamOperator<CoordinatorInput>
```
Yes, it aligns the function with `FileSystemOutputFormat`.

luoyuxia commented on flink-connectors/flink-connector-files/src/main/java/org/apache/flink/connector/file/table/PartitionLoader.java:
```diff
@@ -76,7 +76,8 @@ public PartitionLoader(
 
     /** Load a single partition. */
-    public void loadPartition(LinkedHashMap<String, String> partSpec, List<Path> srcDirs)
+    public void loadPartition(
```
Yes, it must be dir or not simultaneously. There shouldn't be any case that some paths are dir and some are files. If have, it shouldn't be written by Flink.

luoyuxia commented:
@lsyldliu Thanks for contribution. I have addressed your comments.
For the parallelism of `CompactOperator` in aqe , I have created FLINK-29635 to track it.

lsyldliu commented on flink-connectors/flink-connector-files/src/main/java/org/apache/flink/connector/file/table/batch/compact/CompactFileUtils.java:
```diff
@@ -0,0 +1,119 @@
+     * Do Compaction: - Target file exists, do nothing. Otherwise, it'll read the input files and
+     * write the target file to achieve compaction purpose.
+     */
+    public static <T> Path doCompact(
```
```suggestion
    public static <T>  @Nullable Path doCompact(
```

lsyldliu reviewed (commented):
@luoyuxia LGTM overall, I left some minor comments.

luoyuxia commented:
@lsyldliu Thanks for your reviewing. I have addressed your comments.

luoyuxia commented:
@flinkbot run azure

luoyuxia commented:
@wuchong Could you please help merge? Thanks.

wuchong commented:
Thank you all for the contributions.
